paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc

```diff
@@ -0,0 +1,163 @@
+  auto scale_dims_mapping = input_specs[1].dist_attr().dims_mapping();
+  auto bias_dims_mapping = input_specs[2].dist_attr().dims_mapping();
+
+  auto x_dist_attr_src = input_specs[0].dist_attr();
+
+  std::vector<TensorDistAttr> input_dist_attrs;
+  input_dist_attrs.reserve(input_specs.size());
+
+  int begin_norm_axis = ExtractAttr<int>("begin_norm_axis", attrs);
+
+  // Step2.3.2  handle input tensor partial (TODO)
+  VLOG(4) << "LayerNormSPMDRule InferForward Inputs: "
+          << "x shape: [" << str_join(x_shape) << "], x_dims_mapping: ["
+          << str_join(x_dims_mapping) << "]; scale shape: ["
+          << str_join(scale_shape) << "], scale_dims_mapping: ["
+          << str_join(scale_dims_mapping) << "]; bias shape: ["
+          << str_join(bias_shape) << "], bias_dims_mapping: ["
+          << str_join(bias_dims_mapping) << "]; begin_norm_axis: ["
+          << begin_norm_axis << "]; ";
+
+  // step1: build Einsum Notation
+  // ijk,k,k->ijk (x,scale,bias->out, begin_norm_axis=2)
+  // ijkl,z(kl),z(kl)->ijkl (x,scale,bias->out, begin_norm_axis=2, z=kl)
+  std::string x_axes = "";
+  for (auto i = 0; i < x_ndim; ++i) {
+    x_axes += static_cast<char>(static_cast<int>('k') - begin_norm_axis + i);
+  }
+
+  std::string scale_axes;
+  std::string bias_axes;
+  if (x_ndim - begin_norm_axis == 1) {
+    scale_axes = "k";
+    bias_axes = "k";
+  } else {
+    // z = x_axes.substr(begin_norm_axis, x_ndim - begin_norm_axis)
+    scale_axes = "z";
+    bias_axes = "z";
+  }
+  std::string out_axes = x_axes;
+
+  VLOG(4) << "LayerNormSPMDRule build Einsum notation (x,scale,bias->out): ["
+          << x_axes << "," << scale_axes << "," << bias_axes << " --> "
+          << out_axes << "](begin_norm_axis:" << begin_norm_axis
+          << ",z=" << x_axes.substr(begin_norm_axis, x_ndim - begin_norm_axis)
+          << ").";
+
+  // step2: Sharding Propogation
+
+  TensorDistAttr output_dist_attr_dst =
+      CopyTensorDistAttrForOutput(x_dist_attr_src);
+  TensorDistAttr x_dist_attr_dst = CopyTensorDistAttrForOutput(x_dist_attr_src);
+  std::vector<int64_t> out_dims_mapping;
+  out_dims_mapping.reserve(out_axes.size());
+  for (size_t i = 0; i < out_axes.size(); ++i) {
+    if (i < static_cast<size_t>(begin_norm_axis)) {
+      out_dims_mapping.push_back(x_dims_mapping[i]);
+    } else {
+      out_dims_mapping.push_back(-1);
+    }
+  }
+  output_dist_attr_dst.set_dims_mapping(out_dims_mapping);
+  x_dist_attr_dst.set_dims_mapping(out_dims_mapping);
+
+  // step2.3: Merge and get Inputs' New Dims Mapping.
+  input_dist_attrs.emplace_back(x_dist_attr_dst);
+
+  // TODO(zhiqiu): support shardding on scale and bias
+  // Now, apply replicating.
+  input_dist_attrs.emplace_back(ReplicatedOnMesh(input_specs[1].dist_attr()));
+  input_dist_attrs.emplace_back(ReplicatedOnMesh(input_specs[2].dist_attr()));
+
+  // Step2.4.  handle input and out tensor partial
+  std::vector<int64_t> partial_on_dims;
```

Comment:
LayerNorm activation output would not be partial

Reply:
done, thx